src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/OpportunitiesOrchestrator.cs

```diff
@@ -134,12 +189,79 @@ public async Task<ApplyViewModel> GetApplyViewModel(ApplicationRequest request)
                 Sectors = "-",
                 Locations = "-",
                 MoreDetail = "-",
-                ContactName = "-",
-                EmailAddress = "-",
-                WebsiteUrl = "-"
+                ContactName = string.IsNullOrWhiteSpace(contactName) ? "-" : contactName,
+                EmailAddresses = emailAddresses,
+                WebsiteUrl = string.IsNullOrEmpty(application.BusinessWebsite) ? "-" : application.BusinessWebsite,
+            };
+        }
+
+        public async Task<ContactDetailsViewModel> GetContactDetailsViewModel(ContactDetailsRequest contactDetailsRequest)
+        {
+            var getContactDetailsResult = await _opportunitiesService.GetContactDetails(contactDetailsRequest.AccountId, contactDetailsRequest.PledgeId);
+
+            if (getContactDetailsResult == null)
+            {
+                return null;
+            }
+
+            var opportunitySummaryViewModel = GetOpportunitySummaryViewModel(
+                getContactDetailsResult.Sectors.Select(x => x.Id),
+                getContactDetailsResult.JobRoles.Select(x => x.Id),
+                getContactDetailsResult.Levels.Select(x => x.Id),
+                getContactDetailsResult.AllSectors,
+                getContactDetailsResult.AllJobRoles,
+                getContactDetailsResult.AllLevels,
+                getContactDetailsResult.Amount,
+                getContactDetailsResult.IsNamePublic,
+                getContactDetailsResult.DasAccountName,
+                contactDetailsRequest.EncodedPledgeId);
+
+            var cacheItem = await RetrieveCacheItem(contactDetailsRequest.CacheKey);
+
+            var viewModel = new ContactDetailsViewModel()
+            {
+                EncodedAccountId = contactDetailsRequest.EncodedAccountId,
+                EncodedPledgeId = contactDetailsRequest.EncodedPledgeId,
+                FirstName = cacheItem.FirstName,
+                LastName = cacheItem.LastName,
+                EmailAddress = cacheItem.EmailAddresses.First(),
+                AdditionalEmailAddresses = cacheItem.EmailAddresses.Skip(1).ToArray(),
+                BusinessWebsite = cacheItem.BusinessWebsite,
+                DasAccountName = getContactDetailsResult.DasAccountName,
+                OpportunitySummaryViewModel = opportunitySummaryViewModel,
             };
+
+            return viewModel;
+        }
+
+        public async Task UpdateCacheItem(ContactDetailsPostRequest contactDetailsPostRequest)
+        {
+            var cacheItem = await RetrieveCacheItem(contactDetailsPostRequest.CacheKey);
+
+            cacheItem.FirstName = contactDetailsPostRequest.FirstName;
+            cacheItem.LastName = contactDetailsPostRequest.LastName;
+            cacheItem.EmailAddresses = new string[] { contactDetailsPostRequest.EmailAddress }.Concat(contactDetailsPostRequest.AdditionalEmailAddresses.ToArray());
```

Comment:
See earlier comment. Wouldn't it be better to remove the empty entries here, to keep the cache item tidy? Rather than when we get the page, or get the summary page, or when we come to persist it?

Reply:
I have simplified this completely, and as mentioned in a previous comment, I'm only storing non-null values in the cache object now.